utils/time_slots_utils.js

```diff
@@ -0,0 +1,84 @@
+const {models} = require('../db/relations')
+const TimeSlot = models.TimeSlot
+
+const TimeSlotsUtils = {
+
+    getIntersection: function (slots) {
+        let res = []
+        let pointers = new Array(slots.length).fill(0)
+        let l = -1
+        let r = -1
+        while (!isNaN(l) && !isNaN(r) ) {
+            let it_slots = new Array(slots.length).fill(0)
+            for (let i = 0; i < slots.length; i++) {
+                it_slots[i] = slots[i][pointers[i]]
+            }
+            let vals = it_slots.map(slot => slot !== undefined ? slot.startDatetime.getTime() : undefined)
+            l = Math.max(...vals)
+
+            vals = it_slots.map(slot => slot !== undefined ? slot.startDatetime.getTime() + slot.duration * 60000 : undefined)
+            r = Math.min(...vals)
+
+            if (l + 15 <= r) {
+                res.push({startDatetime: new Date(l), duration: (new Date(r) - new Date(l)) / 60000})
+            }
+
+            let valsCopy = [...vals]
+            while (valsCopy.indexOf(r) === valsCopy.lastIndex) {
+                valsCopy = valsCopy.filter(v => valsCopy.indexOf(v) !== r)
+                r = Math.min(...valsCopy)
+            }
+            pointers[vals.indexOf(r)]++
+        }
+        return res
+    },
+
+    slice: function (slot, reservation) {
```

Comment:
Nie wiem, czy nie pokusiłbym się o lepszą nazwę tej funkcji - funkcja slice już istnieje i jest wbudowana, więc jak zobaczysz coś takiego, to może być kicha.